File: frontend/src/utils/workspaceFilePolicy.ts
/**
 * Workspace file-type edit policy for the editor and AI chat.
 *
 * Python / JavaScript workspaces can view any file, but only language sources
 * may be edited or AI-enhanced. Website workspaces are unrestricted.
 */

import type { WorkspaceType } from '@/types';

const PYTHON_EDITABLE = new Set(['.py']);
const JAVASCRIPT_EDITABLE = new Set(['.js', '.jsx', '.ts', '.tsx']);

export const PYTHON_EDIT_MESSAGE =
  'Only Python files can be edited or enhanced in this workspace.';

export const JAVASCRIPT_EDIT_MESSAGE =
  'Only JavaScript-related files (.js, .jsx, .ts, .tsx) can be edited or enhanced in this workspace.';

const EDIT_INTENT_RE =
  /\b(edit|enhance|improve|refactor|fix|update|modify|change|rewrite|implement|add|create|delete|remove|generate|replace)\b/i;

/**
 * Return the lowercase file extension including the leading dot.
 *
 * @param path - File path or name.
 * @returns Extension such as `.py`, or empty string for extensionless / dotfiles.
 */
export function fileExtension(path: string): string {
  const name = path.replace(/\\/g, '/').split('/').pop() || '';
  const dot = name.lastIndexOf('.');
  if (dot <= 0) {
    return '';
  }
  return name.slice(dot).toLowerCase();
}

/**
 * Allowed editable extensions for a workspace, or `null` when unrestricted.
 *
 * @param workspaceType - Active workspace.
 * @returns Set of extensions or null.
 */
export function editableExtensionsFor(
  workspaceType: WorkspaceType | string | undefined,
): Set<string> | null {
  const workspace = (workspaceType || '').toLowerCase();
  if (workspace === 'python') {
    return PYTHON_EDITABLE;
  }
  if (workspace === 'javascript') {
    return JAVASCRIPT_EDITABLE;
  }
  return null;
}

/**
 * User-facing restriction message for a workspace.
 *
 * @param workspaceType - Active workspace.
 * @returns Modal message text.
 */
export function editRestrictionMessage(
  workspaceType: WorkspaceType | string | undefined,
): string {
  const workspace = (workspaceType || '').toLowerCase();
  if (workspace === 'python') {
    return PYTHON_EDIT_MESSAGE;
  }
  if (workspace === 'javascript') {
    return JAVASCRIPT_EDIT_MESSAGE;
  }
  return 'This file type cannot be edited in the current workspace.';
}

/**
 * Whether a path may be edited / AI-enhanced in the workspace.
 *
 * @param workspaceType - Active workspace.
 * @param path - File path.
 * @returns True when edits are allowed.
 */
export function isPathEditable(
  workspaceType: WorkspaceType | string | undefined,
  path: string | undefined | null,
): boolean {
  if (!path) {
    return true;
  }
  const allowed = editableExtensionsFor(workspaceType);
  if (!allowed) {
    return true;
  }
  return allowed.has(fileExtension(path));
}

/**
 * Whether a chat message looks like an edit/enhance request.
 *
 * @param content - User chat text.
 * @returns True when the message implies file mutation.
 */
export function isEditIntentMessage(content: string): boolean {
  return EDIT_INTENT_RE.test(content || '');
}

File: backend/app/utils/workspace_file_policy.py
"""Workspace file-type edit policy.

Python and JavaScript workspaces may contain mixed files for viewing, but only
language-specific extensions may be created, updated, deleted, or AI-enhanced.
Website workspaces remain unrestricted for typical web assets.
"""

from __future__ import annotations

from typing import FrozenSet, Optional

from app.utils.exceptions import ValidationAppError

PYTHON_EDITABLE_EXTENSIONS: FrozenSet[str] = frozenset({".py"})
JAVASCRIPT_EDITABLE_EXTENSIONS: FrozenSet[str] = frozenset({".js", ".jsx", ".ts", ".tsx"})

PYTHON_EDIT_MESSAGE = "Only Python files can be edited or enhanced in this workspace."
JAVASCRIPT_EDIT_MESSAGE = (
    "Only JavaScript-related files (.js, .jsx, .ts, .tsx) can be edited or "
    "enhanced in this workspace."
)


def _extension_of(path: str) -> str:
    name = (path or "").replace("\\", "/").rsplit("/", 1)[-1]
    if "." not in name or name.startswith("."):
        # Dotfiles like `.env` are not editable language sources in restricted workspaces.
        return ""
    return f".{name.rsplit('.', 1)[-1].lower()}"


def editable_extensions_for(workspace_type: str) -> Optional[FrozenSet[str]]:
    """Return allowed extensions for a workspace, or ``None`` when unrestricted.

    Args:
        workspace_type: javascript | python | website.

    Returns:
        Frozen set of extensions including the leading dot, or ``None``.
    """
    workspace = (workspace_type or "").lower().strip()
    if workspace == "python":
        return PYTHON_EDITABLE_EXTENSIONS
    if workspace == "javascript":
        return JAVASCRIPT_EDITABLE_EXTENSIONS
    return None


def edit_restriction_message(workspace_type: str) -> str:
    """User-facing validation message for a restricted workspace.

    Args:
        workspace_type: javascript | python | website.

    Returns:
        Modal / API error message.
    """
    workspace = (workspace_type or "").lower().strip()
    if workspace == "python":
        return PYTHON_EDIT_MESSAGE
    if workspace == "javascript":
        return JAVASCRIPT_EDIT_MESSAGE
    return "This file type cannot be edited in the current workspace."


def is_path_editable(workspace_type: str, path: str) -> bool:
    """Whether ``path`` may be created/updated/deleted in the workspace.

    Args:
        workspace_type: Project workspace type.
        path: Project-relative file path.

    Returns:
        True when editing is allowed.
    """
    allowed = editable_extensions_for(workspace_type)
    if allowed is None:
        return True
    return _extension_of(path) in allowed


def assert_path_editable(workspace_type: str, path: str) -> None:
    """Raise ``ValidationAppError`` when ``path`` is not editable.

    Args:
        workspace_type: Project workspace type.
        path: Project-relative file path.

    Raises:
        ValidationAppError: When the extension is not allowed.
    """
    if not is_path_editable(workspace_type, path):
        raise ValidationAppError(
            edit_restriction_message(workspace_type),
            details={
                "workspace_type": (workspace_type or "").lower().strip(),
                "path": path,
                "error_code": "workspace_file_type_restricted",
            },
        )
